### PR TITLE
add task type controller implementation

### PR DESCRIPTION
**Pull request by Glunky**

### CODE REVIEW

File: src/DashboardService.Business/TaskType/CreateTaskTypeCommand.cs [hunk 1]
```diff
@@ -1,24 +1,63 @@
...
 
-  public CreateTaskTypeCommand(ITaskTypeRepository repository)
+  public CreateTaskTypeCommand(
+    ITaskTypeRepository repository,
+    IAccessValidator accessValidator,
+    ICreateTaskTypeRequestValidator validator,
+    IResponseCreator responseCreator,
+    IDbTaskTypeMapper taskTypeMapper,
+    IHttpContextAccessor httpContextAccessor
+    )
```
Comment (VLiamin) [inline src/DashboardService.Business/TaskType/CreateTaskTypeCommand.cs]: on previous

File: src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+namespace LT.DigitalOffice.DashboardService.Validation.TaskType;
+
+public class EditTaskTypeRequestValidator : ExtendedEditRequestValidator<Guid, EditTaskTypeRequest>, IEditTaskTypeRequestValidator
+{
+  private readonly ITaskTypeRepository _taskTypeRepository;
+
+  public EditTaskTypeRequestValidator(ITaskTypeRepository repository)
+  {
+    _taskTypeRepository = repository;
+    RuleFor(x => x)
```
Comment (VLiamin) [inline src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs:23]: empty line

File: src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    await AddFailureForPropertyIfNotAsync(
+      nameof(EditTaskTypeRequest.Name),
+      x => x == OperationType.Replace,
+      new()
+      {
+        {
+          async x =>
+            !await _taskTypeRepository.NameExistAsync(x.value?.ToString(), taskTypeId),
+          "The department name already exist."
```
Comment (VLiamin) [inline src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs]: tab

File: src/DashboardService/appsettings.Development.json [hunk 4]
```diff
@@ -9,7 +9,7 @@
     }
   },
   "ConnectionStrings": {
-    "SQLConnectionString": "Server=localhost\\sqlexpress;Database=DashboardDB;Trusted_Connection=True;Encrypt=False"
+    "SQLConnectionString": "Server=DESKTOP-M9M79TR\\SQLEXPRESS;Database=DashboardDB;Trusted_Connection=True;Encrypt=False"
```
Comment (Glunky) [inline src/DashboardService/appsettings.Development.json:12]: do not commit

File: src/DashboardService.Business/TaskType/CreateTaskTypeCommand.cs (hunk 1 above)
Comment (Glunky) [reply src/DashboardService.Business/TaskType/CreateTaskTypeCommand.cs]: ++

File: src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs (hunk 2 above)
Comment (Glunky) [reply src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs:23]: ++

File: src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs (hunk 3 above)
Comment (Glunky) [reply src/DashboardService.Validation/TaskType/EditTaskTypeRequestValidator.cs]: ++

File: src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs [hunk 5]
```diff
@@ -1,25 +1,60 @@
...
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+    }
+
+    ValidationResult validationResult = await _validator.ValidateAsync((id, request));
+
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest,
+        validationResult.Errors.Select(vf => vf.ErrorMessage).ToList());
```
Comment (VLiamin) [inline src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: validationResult.Errors is List or ICollection?

Comment (Glunky) [reply src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: List 
image

Comment (VLiamin) [reply src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: So may be change Select(vf => vf.ErrorMessage).ToList() to ConvertAll()?

File: src/DashboardService.Data/TaskTypeRepository.cs [hunk 6]
```diff
@@ -18,28 +21,83 @@ public TaskTypeRepository(IDataProvider provider)
...
+  {
+    DbTaskType dbTaskType = await GetAsync(id, ct);
+
+    if (dbTaskType is null)
+    {
+      return false;
+    }
+
+    _provider.TaskTypes.Remove(dbTaskType);
+    await _provider.SaveAsync();
```
Comment (VLiamin) [inline src/DashboardService.Data/TaskTypeRepository.cs:87]: ct?

File: src/DashboardService.Business/TaskType/GetTaskTypesCommand.cs [hunk 7]
```diff
@@ -1,25 +1,39 @@
...
   
-  public Task<FindResultResponse<TaskTypeInfo>> ExecuteAsync(GetTaskTypesFilter filter)
+  public async Task<FindResultResponse<TaskTypeInfo>> ExecuteAsync(GetTaskTypesFilter filter, CancellationToken ct)
   {
-    throw new System.NotImplementedException();
+    (List<DbTaskType> dbTaskTypes, int totalCount) result = await _taskTypeRepository.GetAllAsync(filter, ct);
+
+    return new()
+    {
+      Body = result.dbTaskTypes.Select(p => _taskTypeInfoMapper.Map(p)).ToList(),
```
Comment (VLiamin) [inline src/DashboardService.Business/TaskType/GetTaskTypesCommand.cs]: May be change p => _taskTypeInfoMapper.Map(p) on _taskTypeInfoMapper.Map?

File: src/DashboardService.Data/TaskTypeRepository.cs (hunk 6 above)
Comment (Glunky) [reply src/DashboardService.Data/TaskTypeRepository.cs:87]: for SaveAsync no overloads with ct as argument

Comment (Glunky) [reply src/DashboardService.Data/TaskTypeRepository.cs:87]: we dont have in IBaseDataProvider overloads for method Task SaveAsync() with ct as argument and i cant use here ct. only if add same method to IDataProvider, but i dont think that this is correct class for such method

File: src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs (hunk 5 above)
Comment (Glunky) [reply src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: ++

File: src/DashboardService.Business/TaskType/GetTaskTypesCommand.cs (hunk 7 above)
Comment (Glunky) [reply src/DashboardService.Business/TaskType/GetTaskTypesCommand.cs]: ++

File: src/DashboardService.Data/TaskTypeRepository.cs (hunk 6 above)
Comment (VLiamin) [reply src/DashboardService.Data/TaskTypeRepository.cs:87]: Yes, I mixed it up with the method SaveChangesAsync of DbContext

File: src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs [hunk 8]
```diff
@@ -1,25 +1,61 @@
...
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+    }
+
+    ValidationResult validationResult = await _validator.ValidateAsync((id, request), ct);
+
+    if (!validationResult.IsValid)
+    {
+      return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.BadRequest,
+        validationResult.Errors.ConvertAll(vf => vf.ErrorMessage));
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: split

File: src/DashboardService.Business/TaskType/RemoveTaskTypeCommand.cs [hunk 9]
```diff
@@ -1,23 +1,53 @@
...
+    if (!await _accessValidator.IsAdminAsync())
+    {
+      _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+    }
+
+    if (!await _taskTypeRepository.RemoveAsync(id, ct))
+    {
+      return _responseCreator.CreateFailureResponse<bool>(
+        HttpStatusCode.NotFound,
+        new() {"Cannot remove task type with given Id"});
```
Comment (SolovyovDmitry) [inline src/DashboardService.Business/TaskType/RemoveTaskTypeCommand.cs]: spaces after and before {}

**Review by SolovyovDmitry — APPROVED** [review]

File: src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs (hunk 8 above)
Comment (Glunky) [reply src/DashboardService.Business/TaskType/EditTaskTypeCommand.cs]: ++

File: src/DashboardService.Business/TaskType/RemoveTaskTypeCommand.cs (hunk 9 above)
Comment (Glunky) [reply src/DashboardService.Business/TaskType/RemoveTaskTypeCommand.cs]: ++

File: src/DashboardService.Data/TaskTypeRepository.cs [hunk 10]
```diff
@@ -18,28 +21,78 @@ public TaskTypeRepository(IDataProvider provider)
...
   {
-    throw new NotImplementedException();
+    DbTaskType dbTaskType = await _provider.TaskTypes.FindAsync(id, ct);
+
+    if (dbTaskType is null)
+    {
+      return false;
+    }
+
+    request.ApplyTo(dbTaskType);
```
Comment (VLiamin) [inline src/DashboardService.Data/TaskTypeRepository.cs:70]: Edit ModifiedBy and ModifiedAtUtc?

**Review by VLiamin — APPROVED** [review]

File: src/DashboardService.Data/TaskTypeRepository.cs (hunk 10 above)
Comment (Glunky) [reply src/DashboardService.Data/TaskTypeRepository.cs:70]: for task type entity - no

**Pull request merged**